pkg/storegateway/bucket_index_postings.go

```diff
@@ -322,3 +324,143 @@ func checkNilPosting(l labels.Label, p index.Postings) index.Postings {
+// worstCaseFetchedDataStrategy uses a fixed estimation about the size of series in the index (tsdb.EstimatedSeriesP99Size).
+type worstCaseFetchedDataStrategy struct {
+	// postingListActualSizeFactor affects how
+	// posting lists are summed together.
+	// This factor is multiplied by the size of posting lists
+	// in the block index.
+	postingListActualSizeFactor float64
+}
+
+func (s worstCaseFetchedDataStrategy) name() string {
+	return fmt.Sprintf("worstCase%0.1f", s.postingListActualSizeFactor)
+}
+
+func (s worstCaseFetchedDataStrategy) selectPostings(groups []postingGroup) (selected, omitted []postingGroup) {
+	const seriesBytesPerPostingByte = tsdb.EstimatedSeriesP99Size / tsdb.BytesPerPostingInAPostingList
+
+	sort.Slice(groups, func(i, j int) bool {
+		return groups[i].totalSize < groups[j].totalSize
+	})
+
+	var minGroupSize int64
+	for _, g := range groups {
+		// The size of each posting list contains 4 bytes with the number of entries.
+		// We shouldn't count these as series.
+		if !g.isSubtract && !(len(g.keys) == 1 && g.keys[0] == allPostingsKey) {
+			minGroupSize = g.totalSize - int64(len(g.keys)*4)
+			break
+		}
+	}
+
+	if minGroupSize == 0 {
+		// This should also cover the case of all postings group. all postings is requested only when there is no
+		// additive group.
+		return groups, nil
+	}
+
+	var (
+		selectedSize               int64
+		atLeastOneAdditiveSelected bool
+		maxSelectedSize            = minGroupSize * seriesBytesPerPostingByte
+	)
+	for i, g := range groups {
+		postingListSize := int64(float64(g.totalSize) * s.postingListActualSizeFactor)
+		if atLeastOneAdditiveSelected && selectedSize+postingListSize > maxSelectedSize {
+			return groups[:i], groups[i:]
+		}
+		selectedSize += postingListSize
+		atLeastOneAdditiveSelected = atLeastOneAdditiveSelected || !g.isSubtract
+	}
+	return groups, nil
+}
+
+// speculativeFetchedDataStrategy selects postings lists in a very similar way to worstCaseFetchedDataStrategy,
+// except it speculates on the size of the actual series after intersecting the selected posting lists.
+// Right now it assumes that each posting list will
+type speculativeFetchedDataStrategy struct{}
+
+func (s speculativeFetchedDataStrategy) name() string {
+	return "speculative"
+}
+
+func (s speculativeFetchedDataStrategy) selectPostings(groups []postingGroup) (selected, omitted []postingGroup) {
+	const (
+		postingsPerByteInPostingList = 4
+		bytesPerSeries               = 512
+
+		seriesBytesPerPostingByte = bytesPerSeries / postingsPerByteInPostingList
+	)
+
+	sort.Slice(groups, func(i, j int) bool {
+		return groups[i].totalSize < groups[j].totalSize
+	})
+
+	var minGroupSize int64
+	for _, g := range groups {
+		// The size of each posting list contains 4 bytes with the number of entries.
+		// We shouldn't count these as series.
+		if !g.isSubtract && !(len(g.keys) == 1 && g.keys[0] == allPostingsKey) {
+			minGroupSize = g.totalSize - int64(len(g.keys)*4)
+			break
+		}
+	}
```

Comment:
[nit] Given this is repeated between the two strategies, you may extract it into a function.